src/plugins/gpgme/gpgme.c

```diff
@@ -0,0 +1,580 @@
+				   "No valid recipients were specified. Please provide at least one valid GPG key.");
+		returnValue = -1;
+		goto cleanup;
+	}
+
+	if (inTestMode (pluginConfig))
+	{
+		encryptFlags |= GPGME_ENCRYPT_ALWAYS_TRUST;
+	}
+
+	ksRewind (data);
+	while ((k = ksNext (data)))
+	{
+		gpgme_data_t input;
+		gpgme_data_t ciphertext;
+		gpgme_encrypt_result_t result;
+		gpgme_invalid_key_t invalidKey;
+		int keyDataType;
+
+		if (!isMarkedForEncryption (k) || isSpecNamespace (k) || isNullValue (k))
+		{
+			continue;
+		}
+
+		// preserve the data type of k (string, binary)
+		keyDataType = keyIsBinary (k);
+		if (!keyIsBinary (k))
+		{
+			err = gpgme_data_new_from_mem (&input, keyString (k), strlen (keyString (k)) + 1, 0);
+		}
+		else
+		{
+			keySetMeta (k, ELEKTRA_GPGME_META_BINARY, "1");
+			err = gpgme_data_new_from_mem (&input, keyValue (k), keyGetValueSize (k), 0);
+		}
+		if (err)
+		{
+			returnValue = -1;
+			ELEKTRA_SET_ERRORF (ELEKTRA_ERROR_GPGME, errorKey, "Internal error: %s", gpgme_strerror (err));
+			goto cleanup;
+		}
+
+		err = gpgme_data_new (&ciphertext);
+		if (err)
+		{
+			returnValue = -1;
+			ELEKTRA_SET_ERRORF (ELEKTRA_ERROR_GPGME, errorKey, "Internal error: %s", gpgme_strerror (err));
+			gpgme_data_release (input);
+			goto cleanup;
+		}
+
+		err = gpgme_op_encrypt (ctx, recipients, encryptFlags, input, ciphertext);
+		if (err)
+		{
+			returnValue = -1;
+			ELEKTRA_SET_ERRORF (ELEKTRA_ERROR_GPGME, errorKey, "Internal error: %s", gpgme_strerror (err));
+			gpgme_data_release (ciphertext);
+			gpgme_data_release (input);
+			goto cleanup;
+		}
+
+		result = gpgme_op_encrypt_result (ctx);
+		invalidKey = result->invalid_recipients;
+		while (invalidKey)
+		{
+			ELEKTRA_ADD_WARNINGF (ELEKTRA_WARNING_GPGME_INVALID_RECIPIENT, errorKey, "Invalid GPG key ID: %s", invalidKey->fpr);
+			invalidKey = invalidKey->next;
+		}
+
+		// update Elektra key to encrypted value
+		if (transferGpgmeDataToElektraKey (ciphertext, k, errorKey, textMode) != 1)
+		{
+			// error description has been set by transferGpgmeDataToElektraKey()
+			returnValue = -1;
+			gpgme_data_release (ciphertext);
+			gpgme_data_release (input);
+			goto cleanup;
+		}
+
+		gpgme_data_release (ciphertext);
+		gpgme_data_release (input);
+	}
+
+cleanup:
+	freeRecipientArray (recipients);
+	gpgme_release (ctx);
+	return returnValue;
+}
+
+/**
+ * @brief Decrypt all Keys in the KeySet "data" that have a meta-key "gpg/encrpyt" set.
+ * @see isMarkedForEncryption (const Key * k)
+ * @param handle holds the plugin configuration.
+ * @param data the KeySet to be decrypted.
+ * @param errorKey will hold and error description if an operation fails.
+ * @retval 1 on success.
+ * @retval -1 on failure. In this case errorKey will provide a description.
+ */
+static int gpgDecrypt (ELEKTRA_UNUSED Plugin * handle, KeySet * data, Key * errorKey)
+{
+	int returnValue = 1; // success
+	Key * k;
+
+	gpgme_ctx_t ctx;
+	gpgme_error_t err;
+
+	err = gpgme_new (&ctx);
```

Comment:
Wouldn't it be cheaper if one context is stored in the handle and used for everything? It would also avoid the goto.